flang/lib/Semantics/resolve-names.cpp

```diff
@@ -2390,7 +2390,7 @@ Symbol &ScopeHandler::MakeHostAssocSymbol(
   // These attributes can be redundantly reapplied without error
   // on the host-associated name, at most once (C815).
   symbol.implicitAttrs() =
-      symbol.attrs() & Attrs{Attr::ASYNCHRONOUS, Attr::VOLATILE};
+      symbol.attrs() & Attrs{Attr::ASYNCHRONOUS, Attr::VOLATILE, Attr::SAVE};
```

Comment:
An explicit `SAVE` statement for a name in a nested scope must create a new symbol.  It is not possible for a `SAVE` statement in a nested scope to apply to a host-associated symbol.

Reply:
I had to make this change to resolve the following issue:
```
subroutine host
  integer :: x = 1
  integer :: y
  common /blk/ y
  equivalence(x,y)
contains
  subroutine inner
    x = y
  end subroutine inner
end subroutine host
```

Without propagating to the host-associated symbol the fact that `SAVE` is implicit, I would get this error:
```
./save.f90:8:5: error: The entity 'x' in COMMON block /blk/ may not have an explicit SAVE attribute
      x = y
      ^
./save.f90:8:9: error: The entity 'y' in COMMON block /blk/ may not have an explicit SAVE attribute
      x = y
          ^
bbc: semantic errors in save.f90
```

As you said, explicit `SAVE` statement in a nested scope will create a new symbol, so this code should not affect this case.

Reply:
Have you tested that an explicit SAVE in the nested scope still creates a new symbol?

If it does, please at least change the comment before your change.

Reply:
For this:
```
subroutine host
  integer :: x = 1
  integer :: y
  common /blk/ y
  equivalence(x,y)
contains
  subroutine inner
    save x
    save y
    x = y
  end subroutine inner
end subroutine host
```

I get this:
```
  Subprogram scope: host size=0 alignment=1 sourceRange=164 bytes
    .F18.0, SAVE (CompilerCreated) size=4 offset=0: ObjectEntity type: INTEGER(4) shape: 1_8:1_8 init:[INTEGER(4)::1_4]
    host (Subroutine): HostAssoc
    inner (Subroutine): Subprogram ()
    x, SAVE size=4 offset=0: ObjectEntity type: INTEGER(4) init:1_4
    y, SAVE size=4 offset=0: ObjectEntity type: INTEGER(4)
    Equivalence Sets: (x,y) (.F18.0,y)
    blk size=4 offset=0: CommonBlockDetails alignment=4: y
    Subprogram scope: inner size=8 alignment=4 sourceRange=57 bytes
      inner (Subroutine): HostAssoc
      x, SAVE (Implicit) size=4 offset=0: ObjectEntity type: REAL(4)
      y, SAVE (Implicit) size=4 offset=4: ObjectEntity type: REAL(4)
```

So the new symbols are created.

I will upload an update shortly.